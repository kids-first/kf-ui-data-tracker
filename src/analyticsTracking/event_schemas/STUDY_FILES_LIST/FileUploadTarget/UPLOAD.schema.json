{
  "$id": "http://kf-ui-data-tracker.kidsfirstdrc.org/src/analyticsTracking/event_schemas/STUDY_FILES_LIST/FileUploadTarget/UPLOAD.schema.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "UPLOAD",
  "description": "",
  "properties": {
    "path": {
      "$ref": "http://kf-ui-data-tracker.kidsfirstdrc.org/src/analyticsTracking/event_schemas/common_defs.schema.json#/definitions/path"
    },
    "route": {
      "$ref": "http://kf-ui-data-tracker.kidsfirstdrc.org/src/analyticsTracking/event_schemas/common_defs.schema.json#/definitions/route"
    },
    "view": {
      "$ref": "http://kf-ui-data-tracker.kidsfirstdrc.org/src/analyticsTracking/event_schemas/common_defs.schema.json#/definitions/view"
    },
    "scope": {
      "$ref": "http://kf-ui-data-tracker.kidsfirstdrc.org/src/analyticsTracking/event_schemas/common_defs.schema.json#/definitions/scope"
    },
    "study": {
      "allOf": [
        {
          "$ref": "http://kf-ui-data-tracker.kidsfirstdrc.org/src/analyticsTracking/event_schemas/common_defs.schema.json#/definitions/study"
        },
        {
          "properties": {
            "files": {
              "type": "number",
              "description": "number of files in the study "
            }
          }
        }
      ]
    },
    "file": {
      "type": "object",
      "description": "file properties from the file metadata on the users local file systemz",
      "properties": {
        "file_name": {
          "type": "string",
          "description": "name of the file on the users local file system"
        },
        "type": {
          "type": "string",
          "description": "MIME type of file form users local file system",
          "examples": ["image/jpg", "text/csv"]
        },
        "size": {
          "type": "integer",
          "description": "size in bites of local file"
        },
        "last_modified": {
          "type": "integer",
          "description": "UTC time that file file was modified on the local system"
        },

        "last_modified_date": {
          "type": ["string", "object"],
          "format": "date-time",
          "description": "time the study document was last modified"
        }
      },
      "required": ["file_name", "type"]
    }
  },
  "required": ["path", "route", "view", "study", "file"],
  "additionalProperties": false
}
