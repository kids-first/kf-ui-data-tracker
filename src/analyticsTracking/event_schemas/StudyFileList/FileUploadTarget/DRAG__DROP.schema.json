{
  "$id": "https://kf-ui-datatracker.kidsfirstdrc.org/StudyFileList/FileUploadTarget/DRAG__DROP.schema.json",
  "$schema": "http://json-schema.org/draft-04/schema#",
  "title": "DRAG__DROP",
  "type": "object",
  "description": "onDrop event for StudyFileList drag and drop FileUploadTarget",
  "allOf": [
    {"$ref": "../../common_defintions.json#/path"},
    {"$ref": "../../common_defintions.json#/route"},
    {"$ref": "../../common_defintions.json#/view"},
    {"$ref": "../../common_defintions.json#/scope"},
    {
      "properties": {
        "study": {
          "allOf": [
            {"$ref": "../../common_defintions.json#/study"},
            {
              "properties": {
                "files": {
                  "type": "number",
                  "description": "number of files in the study "
                }
              }
            }
          ]
        }
      }
    }
  ]
}
