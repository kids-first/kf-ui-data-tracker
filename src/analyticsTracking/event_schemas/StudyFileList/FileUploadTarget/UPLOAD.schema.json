{
  "$id": "https://kf-ui-datatracker.kidsfirstdrc.org/event_schemas/UPLOAD/UPLOAD.schema.json",
  "$schema": "http://json-schema.org/draft-04/schema#",
  "title": "UPLOAD",
  "type": "object",
  "description": "",
  "allOf": [
    {"$ref": "../../common_defintions.json#/path"},
    {"$ref": "../../common_defintions.json#/route"},
    {"$ref": "../../common_defintions.json#/view"},
    {"$ref": "../../common_defintions.json#/scope"},
    {"$ref": "../../common_defintions.json#/file"},
    {
      "properties": {
        "study": {
          "allOf": [
            {"$ref": "../../common_defintions.json#/study"},
            {
              "properties": {
                "files": {
                  "type": "number",
                  "description": "number of files in the study "
                }
              }
            }
          ]
        }
      }
    }
  ]
}
