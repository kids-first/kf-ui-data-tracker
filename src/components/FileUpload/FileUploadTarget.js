import React from 'react';
import classes from 'classnames';
import PropTypes from 'prop-types';

const FileUploadTarget = props => {
  const {
    className,
    error,
    instructions,
    handleSelectedFile,
    handleDragOver,
    handleDragEnter,
    handleDragLeave,
    handleDrop,
  } = props;

  const fileUploadFormClass = classes(
    'upload-target',
    'w-full',
    'p-32',
    'bg-lightGrey',
    'border-grey',
    'border-2',
    'border-dashed',
    'text-center',
    className,
  );

  return (
    <form
      className={fileUploadFormClass}
      onDragOver={handleDragOver}
      onDragEnter={handleDragEnter}
      onDragLeave={handleDragLeave}
      onDrop={handleDrop}
    >
      {instructions ? (
        <p className="max-w-full pt-0 pb-4 mt-0">
          {instructions}
          <br />
          <small className="m-0 p-0">
            <i>or</i>
          </small>
        </p>
      ) : null}
      <input
        type="file"
        className="Button Button--default"
        onChange={handleSelectedFile}
      />
      {error && error.graphQLErrors && (
        <h4 className="text-red px-2">
          Failed to upload: {error.graphQLErrors.map(err => err.message)}
        </h4>
      )}
    </form>
  );
};

FileUploadTarget.propTypes = {
  /** any additional classes that should be added to the container */
  className: PropTypes.string,
  /** if cursor is currently dragging over the component */
  dragging: PropTypes.bool,
  /** Errors from graphQL */
  errors: PropTypes.object,
  /** add helpful instruction about how to drag and drop files */
  instructions: PropTypes.string,
  /** operation when component is dragged over */
  handleDragOver: PropTypes.func.isRequired,
  /** operation when the cursor enters the componen in drag state */
  onDragEnter: PropTypes.func.isRequired,
  /** operation when cursor leaves component in drag state */
  onDragLeave: PropTypes.func.isRequired,
  /** operation when a drag is released on the component */
  handleDrop: PropTypes.func.isRequired,
  /** operations for selected files */
  handleSelectedFile: PropTypes.func.isRequired,
};

FileUploadTarget.defaultProps = {
  className: null,
  instructions: 'Drag and drop Files here',
  dragging: false,
};

/**
 * @component
 */
export default FileUploadTarget;
