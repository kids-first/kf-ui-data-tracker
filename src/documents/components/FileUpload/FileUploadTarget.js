import React from 'react';
import PropTypes from 'prop-types';
import {Button, Header, Icon, Segment, Message} from 'semantic-ui-react';
import {withAnalyticsTracking} from '../../../analyticsTracking';

const FileUploadTarget = props => {
  const {
    error,
    instructions,
    handleSelectedFile,
    handleDragOver,
    handleDragEnter,
    handleDragLeave,
    handleDrop,
    tracking: {
      instrument,
      inheritedEventProps,
      EVENT_CONSTANTS: {DRAG},
    },
  } = props;

  return (
    <form
      onDragOver={handleDragOver}
      onDragEnter={instrument(DRAG.ENTER, handleDragEnter)}
      onDragLeave={instrument(DRAG.LEAVE, handleDragLeave)}
      onDrop={instrument(DRAG.DROP, handleDrop)}
    >
      <Segment piled placeholder>
        {instructions ? (
          <Header icon>
            <Icon name="file outline" />
            {instructions}
            <br />
            <small>or</small>
          </Header>
        ) : null}
        <br />
        <Button icon primary labelPosition="left" as="label" htmlFor="file">
          <Icon name="file outline" />
          Choose a file
        </Button>
        <input
          hidden
          multiple
          id="file"
          type="file"
          onChange={handleSelectedFile}
        />
        {error && error.graphQLErrors && (
          <Message
            negative
            icon="warning circle"
            header="Failed to upload:"
            content={error.graphQLErrors.join(', ')}
          />
        )}
      </Segment>
    </form>
  );
};

FileUploadTarget.propTypes = {
  /** Errors from graphQL */
  error: PropTypes.object,
  /** add helpful instruction about how to drag and drop files */
  instructions: PropTypes.string,
  /** operation when component is dragged over */
  handleDragOver: PropTypes.func.isRequired,
  /** operation when the cursor enters the componen in drag state */
  handleDragEnter: PropTypes.func.isRequired,
  /** operation when cursor leaves component in drag state */
  handleDragLeave: PropTypes.func.isRequired,
  /** operation when a drag is released on the component */
  handleDrop: PropTypes.func.isRequired,
  /** operations for selected files */
  handleSelectedFile: PropTypes.func.isRequired,
};

FileUploadTarget.defaultProps = {
  instructions: 'Drag and drop Files here',
};

/**
 * @component
 */
export default withAnalyticsTracking(FileUploadTarget);
