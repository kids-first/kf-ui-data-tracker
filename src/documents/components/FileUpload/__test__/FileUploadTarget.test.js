import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {render} from 'react-testing-library';
import FileUploadTarget from '../FileUploadTarget';

it('renders correctly', () => {
  const tree = render(
    <MemoryRouter>
      <FileUploadTarget
        handleSelectedFile={jest.fn()}
        handleDragOver={jest.fn()}
        handleDragEnter={jest.fn()}
        handleDragLeave={jest.fn()}
        handleDrop={jest.fn()}
      />
    </MemoryRouter>,
  );
  expect(tree.container).toMatchSnapshot();
});

it('renders instructions', () => {
  const tree = render(
    <MemoryRouter>
      <FileUploadTarget
        instructions="Drag files here"
        handleSelectedFile={jest.fn()}
        handleDragOver={jest.fn()}
        handleDragEnter={jest.fn()}
        handleDragLeave={jest.fn()}
        handleDrop={jest.fn()}
      />
    </MemoryRouter>,
  );
  expect(tree.container).toMatchSnapshot();
  expect(tree.queryByText('Drag files here')).not.toBeNull();
});

it('renders errors', () => {
  const tree = render(
    <MemoryRouter>
      <FileUploadTarget
        error={{graphQLErrors: [{message: 'something went wrong'}]}}
        handleSelectedFile={jest.fn()}
        handleDragOver={jest.fn()}
        handleDragEnter={jest.fn()}
        handleDragLeave={jest.fn()}
        handleDrop={jest.fn()}
      />
    </MemoryRouter>,
  );
  expect(tree.container).toMatchSnapshot();
  expect(tree.queryByText(/Failed to upload/)).not.toBeNull();
});
