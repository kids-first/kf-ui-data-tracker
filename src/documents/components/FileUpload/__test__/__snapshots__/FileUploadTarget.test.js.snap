// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <form>
    <div
      class="ui placeholder piled segment"
    >
      <div
        class="ui icon header mb-0"
      >
        <i
          aria-hidden="true"
          class="file outline icon"
        />
        Drag and drop Files here
        <br />
        <small>
          or
        </small>
      </div>
      <br />
      <label
        class="ui icon primary left labeled button"
        for="file"
        role="button"
      >
        <i
          aria-hidden="true"
          class="file outline icon"
        />
        Choose a file
      </label>
      <input
        hidden=""
        id="file"
        multiple=""
        type="file"
      />
    </div>
  </form>
</div>
`;

exports[`renders errors 1`] = `
<div>
  <form>
    <div
      class="ui placeholder piled segment"
    >
      <div
        class="ui icon header mb-0"
      >
        <i
          aria-hidden="true"
          class="file outline icon"
        />
        Drag and drop Files here
        <br />
        <small>
          or
        </small>
      </div>
      <br />
      <label
        class="ui icon primary left labeled button"
        for="file"
        role="button"
      >
        <i
          aria-hidden="true"
          class="file outline icon"
        />
        Choose a file
      </label>
      <input
        hidden=""
        id="file"
        multiple=""
        type="file"
      />
      <div
        class="ui icon negative message"
      >
        <i
          aria-hidden="true"
          class="warning circle icon"
        />
        <div
          class="content"
        >
          <div
            class="header"
          >
            Failed to upload:
          </div>
          <p>
            [object Object]
          </p>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`renders instructions 1`] = `
<div>
  <form>
    <div
      class="ui placeholder piled segment"
    >
      <div
        class="ui icon header mb-0"
      >
        <i
          aria-hidden="true"
          class="file outline icon"
        />
        Drag files here
        <br />
        <small>
          or
        </small>
      </div>
      <br />
      <label
        class="ui icon primary left labeled button"
        for="file"
        role="button"
      >
        <i
          aria-hidden="true"
          class="file outline icon"
        />
        Choose a file
      </label>
      <input
        hidden=""
        id="file"
        multiple=""
        type="file"
      />
    </div>
  </form>
</div>
`;
